File: src/stride_server/routes/pbs.py
"""``GET /api/{user}/pbs`` — personal bests auto-detection."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel

from stride_core.pb_records import (
    ACTIVITY_DISTANCE_TOLERANCE_M,
    DISTANCE_ORDER,
    best_effort_candidates_for_activity,
    load_personal_bests,
)
from stride_core.pb_records import _normalise_date as _core_normalise_date

from ..deps import get_db

router = APIRouter()

# ── Distance tolerance table (metres) ────────────────────────────────────────

DISTANCE_TOLERANCE = ACTIVITY_DISTANCE_TOLERANCE_M

# Canonical ordering for the response list.
_DISTANCE_ORDER = DISTANCE_ORDER


# ── Response schema ───────────────────────────────────────────────────────────


class PBHistoryPoint(BaseModel):
    date: str
    best_so_far_sec: float
    label_id: str | None = None
    source: str | None = None
    segment_start_s: float | None = None
    segment_end_s: float | None = None


class PBEntry(BaseModel):
    distance: str
    race_type: str | None = None
    pb_time_sec: float
    achieved_at: str
    label_id: str
    source: str | None = None
    name: str | None = None
    segment_start_s: float | None = None
    segment_end_s: float | None = None
    history: list[PBHistoryPoint]


class PBsResponse(BaseModel):
    user_id: str
    computed_at: str
    pbs: list[PBEntry]


# ── Detection logic ───────────────────────────────────────────────────────────


def _detect_pbs(
    rows: list[Any],
) -> dict[str, dict]:
    """Compatibility wrapper for old row-only callers.

    Route and coach code use ``load_personal_bests(db)`` so they can scan
    activity timeseries. This wrapper keeps legacy tests/imports working with
    activity-level fallback only.
    """
    running_best: dict[str, float] = {}
    result: dict[str, dict] = {}
    history_by_distance: dict[str, list[dict[str, Any]]] = {}

    for row in rows:
        for candidate in best_effort_candidates_for_activity(
            _NoTimeseriesDb(), row, include_activity_fallback=True,
        ):
            dist = candidate.distance
            if dist in running_best and candidate.duration_s >= running_best[dist]:
                continue
            running_best[dist] = candidate.duration_s
            history = history_by_distance.setdefault(dist, [])
            history.append(candidate.history_point())
            result[dist] = candidate.pb_entry(history)

    return result


class _NoTimeseriesDb:
    def fetch_timeseries(self, _label_id: str) -> list[Any]:
        return []


def _normalise_date(raw: str) -> str:
    return _core_normalise_date(raw)


# ── Route ─────────────────────────────────────────────────────────────────────


@router.get("/api/{user}/pbs", response_model=PBsResponse)
def get_pbs(user: str) -> PBsResponse:
    """Return best-effort PBs for 1K, 3K, 5K, 10K, HM, and FM.

    Reads the persisted ``personal_bests`` table (populated post-sync) instead of
    recomputing the ~7s best-effort scan per request. ``load_personal_bests``
    self-heals when the table was never scanned (idempotent; not guarded
    in-process), and records PB-less users so they aren't re-scanned every call.
    """
    db = get_db(user)
    try:
        pb_map = load_personal_bests(db)
    finally:
        db.close()

    pbs: list[PBEntry] = []
    for dist in _DISTANCE_ORDER:
        if dist not in pb_map:
            continue
        entry = pb_map[dist]
        pbs.append(PBEntry(**entry))

    return PBsResponse(
        user_id=user,
        computed_at=datetime.now(timezone.utc).isoformat(),
        pbs=pbs,
    )

File: src/go/internal/api/pbs.go
// pbs.go is the personal-best read surface: a sibling registrar that shadows
// the FastAPI route in stride_server/routes/pbs.py
//
//   - GET /api/{user}/pbs          (best-effort PBs for 1K/3K/5K/10K/HM/FM)
//
// It reads the persisted personal_bests table (populated post-sync) instead of
// recomputing the ~7s best-effort scan per request, emits the cached entry_json
// (history progression + segment offsets) in the canonical DISTANCE_ORDER, and
// reuses the two auth tiers. The PB detector itself is internal/compute/pb; this
// route is a thin read-through of the storage table.
package api

import (
	"context"
	"encoding/json"
	"net/http"
	"sort"
	"time"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"github.com/zhaochy1990/stride/internal/compute/pb"
	"github.com/zhaochy1990/stride/internal/logging"
	"github.com/zhaochy1990/stride/internal/storage"
)

// ─────────────────────────────────────────────────────────────────────────────
// Dependency. A narrow read port so the api package stays free of GORM.
// Satisfied by *storage.Store.
// ─────────────────────────────────────────────────────────────────────────────

// PBStore is the read surface the pbs endpoint needs.
type PBStore interface {
	PersonalBests(ctx context.Context, userID string) ([]storage.PersonalBest, error)
}

// ─────────────────────────────────────────────────────────────────────────────
// Registrar
// ─────────────────────────────────────────────────────────────────────────────

type pbsRoutes struct {
	store PBStore
	log   *zap.Logger
}

func newPbsRoutes(store PBStore, log *zap.Logger) *pbsRoutes {
	if log == nil {
		log = logging.Default()
	}
	return &pbsRoutes{store: store, log: log}
}

func (p *pbsRoutes) register(rg *gin.RouterGroup) {
	rg.GET("/api/:user/pbs", p.get)
}

// ─────────────────────────────────────────────────────────────────────────────
// Ordering — mirrors stride_core.pb_records.DISTANCE_ORDER.
// ─────────────────────────────────────────────────────────────────────────────

var pbDistanceOrder = []string{"1K", "3K", "5K", "10K", "HM", "FM"}

var pbDistanceRank = func() map[string]int {
	m := make(map[string]int, len(pbDistanceOrder))
	for i, d := range pbDistanceOrder {
		m[d] = i
	}
	return m
}()

// ─────────────────────────────────────────────────────────────────────────────
// DTOs — mirror pbs.PBEntry / PBsResponse. History is passed through as raw
// maps (entry_json already serialises the best-so-far progression) so the wire
// JSON keys are byte-identical to the Python producer.
// ─────────────────────────────────────────────────────────────────────────────

type pbEntryDTO struct {
	Distance      string           `json:"distance"`
	RaceType      string           `json:"race_type"`
	PBTimeSec     float64          `json:"pb_time_sec"`
	AchievedAt    string           `json:"achieved_at"`
	LabelID       string           `json:"label_id"`
	Source        string           `json:"source"`
	Name          *string          `json:"name"`
	SegmentStartS *float64         `json:"segment_start_s,omitempty"`
	SegmentEndS   *float64         `json:"segment_end_s,omitempty"`
	History       []map[string]any `json:"history"`
}

type pbsResponse struct {
	UserID     string       `json:"user_id"`
	ComputedAt string       `json:"computed_at"`
	PBs        []pbEntryDTO `json:"pbs"`
}

// ─────────────────────────────────────────────────────────────────────────────
// Handler
// ─────────────────────────────────────────────────────────────────────────────

// get returns the user's cached personal bests, ordered 1K→FM.
//
//	@Summary		Personal bests
//	@Description	Returns the best-effort personal bests for 1K, 3K, 5K, 10K, HM and FM, read from the persisted personal_bests table (populated post-sync). Only distances present are returned, in canonical order. A user caller may only read their own data; an internal caller may read any user.
//	@Tags			metrics
//	@Produce		json
//	@Param			user	path		string	true	"User id (JWT sub)"
//	@Success		200		{object}	pbsResponse
//	@Failure		401		{object}	errorResponse
//	@Failure		403		{object}	errorResponse
//	@Failure		500		{object}	errorResponse
//	@Security		InternalToken
//	@Security		BearerAuth
//	@Router			/api/{user}/pbs [get]
func (p *pbsRoutes) get(c *gin.Context) {
	user := c.Param("user")
	if !authorizeUser(c, user) {
		return
	}
	ctx := c.Request.Context()

	rows, err := p.store.PersonalBests(ctx, user)
	if err != nil {
		p.log.Error("pbs: read failed", zapErr(err))
		c.JSON(http.StatusInternalServerError, errorResponse{Error: "internal error"})
		return
	}

	// Order by canonical distance (storage returns distance-alphabetical, e.g.
	// HM/FM after 1K..10K would be wrong). Mirrors the Python route iterating
	// _DISTANCE_ORDER.
	sort.Slice(rows, func(i, j int) bool {
		return pbDistanceRank[rows[i].Distance] < pbDistanceRank[rows[j].Distance]
	})

	pbs := make([]pbEntryDTO, 0, len(rows))
	for _, row := range rows {
		var entry pb.Entry
		if err := json.Unmarshal([]byte(row.EntryJSON), &entry); err != nil {
			p.log.Error("pbs: entry_json parse failed", zapErr(err))
			c.JSON(http.StatusInternalServerError, errorResponse{Error: "internal error"})
			return
		}
		pbs = append(pbs, pbEntryDTO{
			Distance:      entry.Distance,
			RaceType:      entry.RaceType,
			PBTimeSec:     entry.PBTimeSec,
			AchievedAt:    entry.AchievedAt,
			LabelID:       entry.LabelID,
			Source:        entry.Source,
			Name:          entry.Name,
			SegmentStartS: entry.SegmentStartS,
			SegmentEndS:   entry.SegmentEndS,
			History:       entry.History,
		})
	}

	c.JSON(http.StatusOK, pbsResponse{
		UserID:     user,
		ComputedAt: time.Now().UTC().Format(time.RFC3339),
		PBs:        pbs,
	})
}
